### DISCUSSION TITLE
Questions about usage of @Transactional on protected and package-private methods + for self-invocation

### DISCUSSION THREAD

**Discussion by omasseau** [quarkusio/quarkus#42049] [Q&A] [answered]
Hello,

I have two questions about `@Transactional` usage in Quarkus.
I cannot find my answer in the documentation : https://quarkus.io/guides/transaction
I also found this thread on SO, but answers are contradicting each others : https://stackoverflow.com/questions/68180747/does-quarkus-transactional-need-public-methods-like-spring-aop

1)  Does the `@Transactional` annotation work on protected and package-private methods ?

Example : 

```
@ApplicationScoped
public class MyService {

    @Transactional
    void method1() {
    }

    @Transactional
    protected void method2() {
    }
}
```

2) Is the `@Transactional` annotation applied for self-invocation when calling a transactional method from the same service ?

Example:

```
@ApplicationScoped
public class MyService {
    
    @Transactional
    public void methodA() {
        // Do something

        methodB(); // Requires a new transaction
    }

    @Transactional(REQUIRES_NEW)
    public void methodB() {
    }
}
```

I know that in traditional Java EE we previously needed to do this : 

```
@ApplicationScoped
public class MyService {

    @Inject
    MyService self;

    @Transactional
    public void methodA() {
        // Do something

        self.methodB(); // Requires a new transaction
    }

    @Transactional(REQUIRES_NEW)
    public void methodB() {
    }
}
```

Is this self injection necessary also in Quarkus or is it not needed ?

(Note: Would be great to add these infos in the documentation)

Thanks :)

**Comment by RemigiuszDudek** [comment]
The part about self invocation is covered in quarkus [cdi documentation](https://quarkus.io/guides/cdi-reference#intercepted-self-invocation). The part about discouraging from using private members is covered in [chapter about private members of cdi documentation](https://quarkus.io/guides/cdi-reference#native-executables-and-private-members)

**Comment by mkouba** [comment, marked as answer]
> 1. Does the `@Transactional` annotation work on protected and package-private methods ?

Yes, it does. However, `private` methods are not intercepted. There's also the [`quarkus.arc.fail-on-intercepted-private-method`](https://quarkus.io/guides/cdi-reference#quarkus-arc_quarkus-arc-fail-on-intercepted-private-method) config property - if enabled (the default), the build fails if a `private` method that is neither an observer nor a producer, is annotated with an interceptor binding.

> 2. Is the @Transactional annotation applied for self-invocation when calling a transactional method from the same service ?

@RemigiuszDudek is right. In Quarkus, intercepted self-invocation works but it's a non-standard feature because CDI does not define this behavior at all.
